### ISSUE TITLE
Progress/status reporting and nested functions

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#1884]
Recently, there was a suggestion to implement progress and status reporting for more functions. 

However, it seems to me that with the current progress reporting mechanism this can get cumbersome when a progress-reporting function `A` calls another progress-reporting function `B`.  One example of such a function could be `layout_kamada_kawai`, which calls `shortest_paths` functions. Without intervention, what happens is that `A` updates the progress to some level less than 100, then `B` overtakes and goes from 0 to 100 again, then `A` takes back control and continues where it left off. The progress bar will be unhelpfully jumping around.

Of course, this is fixable by temporarily disabling the progress handler before calling `B`, but that is a lot of work to implement such a tiny feature.

Can we come up with a better solution for 0.10, where we break a lot of API anyway?

Could we have some way for function `A` to indicate the min and max progress values to be used by `B`? For example (without thinking this completely through):

```
A() {
    // do stuff
    IGRAPH_PROGRESS(10);
    // now we prepare to call B(). When it finishes we want progress to be at 20.
    IGRAPH_PROGRESS_PUSH(20);
    B();
    IGRAPH_PROGRESS_POP();
    // continue incrementing progress within A
}
```

**Comment by ntamas** [comment]
This is something I've had in mind for a while. The stack-based approach sounds good; each entry in the stack would have a `weight` value and ` current` value. `IGRAPH_PROGRESS()` would modify the `current` value, and for sake of simplicity, we could keep `current` as a float in the range [0; 1]. `weight` would express the weight of the task relative to the task immediately below it in the stack. The effective progress would be `100% * \sum_{i=1}^n ( \prod_{j=1}^i weight[j] ) * current[i]` (1-based indexing). Pushing to the stack would push a new entry with a `weight` equal to the number given by the user, and `current` would be set to 0. Popping an element off the stack would increase `current` of the task below the popped-off task with the `weight` of the popped-off task.

**Comment by ntamas** [comment]
The stack probably also needs a message associated to each item, and frontends supporting a single progress bar only would show the bottom-most message (or the top-most one?). Frontends supporting multiple progress bars may recover the exact progress of all tasks from the stack and may present them as stacked progress bars; note that this needs an update to the progress handler interface as well.

**Comment by szhorvat** [comment]
Yes, it would be nice to have a set of nested messages / statuses. It also applies to the status handler (instead of progress handler).

I can see a front end displaying something like:

```
Laying out components. Kamada-Kawai layout. Distance matrix.
```

as the per-component layout calls Kamada-Kwai, which then calls the distance matrix function.

**Comment by ntamas** [comment]
IMHO the status handler is redundant; we can simply have progress handlers where the status of each task may be "indeterminate", and this effectively replicates the functionality of the status handler.

**Comment by szhorvat** [comment]
Another idea here is to do away with the status handler completely (`igraph_status`) and use only the progress handler. If a function cannot provide progress information, only a status, it can set the progress value to NaN.

 - Advantages: A well-integrated system with consistent nested status reporting.
 - Disadvantages: NaN needs special handling to not mess up nested progress reporting.

----

Finally, if we would end up breaking the API completely anyway, let's please move to using the range 0..1 instead of 0..100.

**Comment by szhorvat** [comment]
> IMHO the status handler is redundant; we can simply have progress handlers where the status of each task may be "indeterminate", and this effectively replicates the functionality of the status handler.

Beat me by half a second :-)

**Comment by GroteGnoom** [comment]
All the statusf and STATUSF's now seems to be used as a printf where the printf can be replaced by another function. In general that can be nice as a logging system, but we use it only in very few functions, and I see nothing specific about the functions where statusf is used.

If the plan is to replace these status handlers, how to proceed? It would clean up some code.

**Comment by ntamas** [comment]
I think the original intention for status handlers was to use them in place of progress handlers in cases when it is hard to add a percentage to the operation (because we don't know how much time things will take). High-level interfaces can attach status handlers to igraph. The R interface converts these to ordinary `message()` calls that print stuff to stdout if the `verbose` option of the package is set to `TRUE`. `tkigraph` (a simple GUI-based graph editor in R) shows these messages in the bottom status bar of the window. The Python interface does nothing by default, but you can add a status handler with `igraph.set_status_handler()`, which would then be called with every single message that the C core attempts to print.

If we decide to get rid of the status handler completely, I'd keep `IGRAPH_STATUSF` anyway as a shorthand notation for reporting progress for an item with an indeterminate progress. We would need to get rid of `igraph_set_status_handler()` instead - but IMHO the first step should be to implement proper nested progress handling, otherwise igraph functions using progress reporting that _also_ call other functions that use the status handler would misbehave as the status handler would overwrite the progress information of the calling function.

**Comment by szhorvat** [comment]
> If we decide to get rid of the status handler completely, I'd keep `IGRAPH_STATUSF` anyway as a shorthand notation for reporting progress for an item with an indeterminate progress.

👍 

----

However, I just wanted to point out that status messages are a natural fit for simple printing line by line:

```
Initializing foo ...
Calculating bar ...
```

Progress information is not. It's fine for showing in a status bar or other GUI element, but line-by-line printing is probably not a good idea, as eventually there'll be an algorithm which would flood the console with alternating messages.

**Comment by ntamas** [comment]
The original idea behind status handlers was something similar to the `tkigraph` use-case, i.e. let GUIs wrapping igraph be notified about what igraph is doing at the moment if no progress information is available. I think it makes most sense for "composite" algorithms that tend to take a long time but consist of one or more smaller steps so you can at least tell the user which discrete step you are doing even if you cannot estimate how much time it will take to finish the step. Or, think long-running stochastic optimizations where you could inform the user about some quality measure of the best solution found so far so the user can decide whether the result is "good enough". There aren't too many such functions in igraph yet, and even the ones that could benefit from such a status report (e.g., hierarchical random graph fitting) do not use this, but in theory it's possible.

So I believe there's utility to these functions and it does not seem to cause a maintenance burden, all that we'd need to do is to use the status handlers more extensively where it makes sense, and to wire them up nicely to higher-level interfaces. E.g., Jupyter notebooks could probably also make use of status messages as well as multi-level progress bars.

**Comment by szhorvat** [comment]
This sort of use is replaceable with `IGRAPH_PROGRESS` with a clear message + NaN value (assuming a future design that handles NaN).

**Comment by szhorvat** [comment]
This is a big project that requires breaking changes, re-targeting for 2.0.
